### ISSUE TITLE
Appium test failed - java.lang.NullPointerException

### ISSUE THREAD

**Issue by shubzinator** [appium/appium#6580]
[TestNG] Running:
  /Users/mohammedshuheb/Library/Caches/IntelliJIdea2016.1/temp-testng-customsuite.xml

java.lang.NullPointerException
	at FirstTest.Test1(FirstTest.java:31)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:639)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:816)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1124)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:774)
	at org.testng.TestRunner.run(TestRunner.java:624)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:359)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:354)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:312)
	at org.testng.SuiteRunner.run(SuiteRunner.java:261)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1215)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
	at org.testng.TestNG.run(TestNG.java:1048)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:74)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:121)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)

===============================================
Default Suite
Total tests run: 1, Failures: 1, Skips: 0
===============================================

Process finished with exit code 0

**Comment by imurchie** [comment]
What is on line 31 of `FirstTest.java`?

This seems like a java client issue (see https://github.com/appium/java-client), but you could provide the server logs as a [gist](https://gist.github.com) here so we could see what was happening.

**Comment by shubzinator** [comment]
This is line 31 and 32:

`       WebElement button1 = driver.findElementById("setUpNewCocoon");
        button1.click();`

**Comment by shubzinator** [comment]
I do not have server logs as it's not running correctly.

**Comment by imurchie** [comment]
Is `driver` initialized? The error makes it seem like it is `null`.

**Comment by shubzinator** [comment]
driver is initialised.

**Comment by imurchie** [comment]
If the driver is initialized and a session is started, then there ought to be some server logs.

**Comment by shubzinator** [comment]
This is my class:

`import io.appium.java_client.AppiumDriver;
import io.appium.java_client.android.AndroidDriver;
import org.junit.After;
import org.junit.Before;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.testng.annotations.Test;
import java.net.URL;

public class FirstTest {

    AppiumDriver driver;

    @Before
    public void setup() throws Exception{
        DesiredCapabilities capabilities = new DesiredCapabilities();
        capabilities.setCapability("deviceName", "AndroidTestDevice");

        driver = new AndroidDriver(new URL("127.0.0.1:4723/wd/hub"), capabilities);

    }

    @After
    public void tearDown() throws Exception{
        driver.quit();
    }

    @Test
    public void Test1() throws Exception{
        WebElement button1 = driver.findElementById("setUpNewCocoon");
        button1.click();
    }
}`

**Comment by shubzinator** [comment]
There are NO logs that I can see.

**Comment by imurchie** [comment]
You must be running the Appium server somewhere, though? When you instantiate the `AndroidDriver` it is connecting to a server running on localhost and port 4723. Wherever this is running is where the logs would be.

My guess is that it is _not_ getting instantiated, which leads to the `NullPointerException`. Can you try printing the `driver` in `Test1`?

**Comment by shubzinator** [comment]
It prints 'null'.

**Comment by shubzinator** [comment]
This is all that is showing on the server:

Launching Appium with command: export ANDROID_HOME="/Users/mohammedshuheb/Library/Android/sdk"; '/Applications/Appium.app/Contents/Resources/node/bin/node' appium/build/lib/main.js --address "127.0.01" --debug-log-spacing --automation-name "Appium" --platform-name "Android" --platform-version "5.1" --app "/Users/mohammedshuheb/Downloads/Cocoon_1.0.1091.apk" --device-name "AndroidTestDevice"

[Appium] Welcome to Appium v1.5.3

[Appium] Non-default server args:

[Appium]   address: '127.0.01'

[Appium]   debugLogSpacing: true
[Appium]   platformName: 'Android'
[Appium]   platformVersion: '5.1'
[Appium]   automationName: 'Appium'
[Appium]   deviceName: 'AndroidTestDevice'
[Appium]   app: '/Users/mohammedshuheb/Downloads/Cocoon_1.0.1091.apk'
[Appium] Deprecated server args:
[Appium]   --platform-name => --default-capabilities '{"platformName":"Android"}'
[Appium]   --platform-version => --default-capabilities '{"platformVersion":"5.1"}'
[Appium]   --automation-name => --default-capabilities '{"automationName":"Appium"}'
[Appium]   --device-name => --default-capabilities '{"deviceName":"AndroidTestDevice"}'
[Appium]   --app => --default-capabilities '{"app":"/Users/mohammedshuheb/Downloads/Cocoon_1.0.1091.apk"}'
[Appium] Default capabilities, which will be added to each request unless overridden by desired capabilities:

[Appium]   platformName: 'Android'

[Appium]   platformVersion: '5.1'
[Appium]   automationName: 'Appium'
[Appium]   deviceName: 'AndroidTestDevice'
[Appium]   app: '/Users/mohammedshuheb/Downloads/Cocoon_1.0.1091.apk'

[Appium] Appium REST http interface listener started on 127.0.01:4723

[HTTP] --> GET /wd/hub/status {}

[MJSONWP] Calling AppiumDriver.getStatus() with args: []

[MJSONWP] Responding to client with driver.getStatus() result: {"build":{"version":"1.5.3"...

[HTTP] <-- GET /wd/hub/status 200 28 ms - 83 

[HTTP] --> GET /wd/hub/status {}

[MJSONWP] Calling AppiumDriver.getStatus() with args: []

[MJSONWP] Responding to client with driver.getStatus() result: {"build":{"version":"1.5.3"...
[HTTP] <-- GET /wd/hub/status 200 13 ms - 83 

[HTTP] --> GET /wd/hub/status {}

[MJSONWP] Calling AppiumDriver.getStatus() with args: []

[MJSONWP] Responding to client with driver.getStatus() result: {"build":{"version":"1.5.3"...
[HTTP] <-- GET /wd/hub/status 200 11 ms - 83 

[HTTP] --> GET /wd/hub/status {}

[MJSONWP] Calling AppiumDriver.getStatus() with args: []

[MJSONWP] Responding to client with driver.getStatus() result: {"build":{"version":"1.5.3"...
[HTTP] <-- GET /wd/hub/status 200 10 ms - 83

**Comment by shubzinator** [comment]
It is listening, but doesn't pick anything up when I run the test.

**Comment by imurchie** [comment]
Yes. It is something to do with how the test runner is initializing the test. I would ask on https://discuss.appium.io/ and perhaps check with the folks working on the Java client to see if they can help.

**Comment by polepalli-12** [comment]
Is this issue resoved iam also facing same problem

**Comment by shabana-parveen** [comment]
@imurchie , selenium latest version 3.5.2 throwing NullPointerExpection for all WebElements. 
I try to resolve this issue in my whole day, then I just downgrade the selenium version with 3.4.0, Now its working. Could you please guide me why the selenium latest version is not supporting the WebElements with the appium?

I tried all elements with click(), clear(), sendkeys() and getText(), getAttribute() etc. 

Please refer below exception log:
java.lang.NullPointerException at org.openqa.selenium.remote.RemoteWebElement.execute(RemoteWebElement.java:275) at org.openqa.selenium.remote.RemoteWebElement.clear(RemoteWebElement.java:117)
----
---

**Comment by imurchie** [comment]
@shabana-parveen The "why" is that the latest Selenium client speaks a slightly different version of Selenium. The Appium Java client is working on getting the changes in to be able to handle things, but it is a juggle (Selenium updates things, and we then have to react to that and update things, in between which it is up in the air whether they will work or not).

**Comment by Amar-Incture** [comment]
I am also facing the same issue. I am using Selenium version as 3.5.2, Appium Desktop version 1.2.0 beta1. The java client used is 5.0.0 Beta 9.
Below is the console log that I get upon executing:

PASSED: setUp
FAILED: operations
java.lang.NullPointerException
	at appium2.Test123.operations(Test123.java:35)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:108)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:661)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:869)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1193)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:126)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:744)
	at org.testng.TestRunner.run(TestRunner.java:602)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:380)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:375)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1301)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1226)
	at org.testng.TestNG.runSuites(TestNG.java:1144)
	at org.testng.TestNG.run(TestNG.java:1115)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:132)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:230)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:76)

My code is as below:

	import java.io.File;
	import java.net.MalformedURLException;
	import java.net.URL;

	import org.openqa.selenium.By;
	import org.openqa.selenium.remote.DesiredCapabilities;
	import org.openqa.selenium.remote.RemoteWebDriver;
	import org.testng.annotations.Test;

	import io.appium.java_client.AppiumDriver;
	import io.appium.java_client.android.AndroidDriver;

	public class Test123 {
		AppiumDriver driver;
		@Test
		public void setUp() throws MalformedURLException{
			File file = new File("/Users/vn0f02x/Downloads/Walmart_v17.13_apkpure.com.apk");
			DesiredCapabilities capabilities = new DesiredCapabilities();
			capabilities.setCapability("device","Android");				
			capabilities.setCapability("app",file.getAbsolutePath());
			capabilities.setCapability("deviceName","ce11160bd19c2d0205");
			capabilities.setCapability("platformName","Android");				
			capabilities.setCapability("platformVersion","7.0");
			capabilities.setCapability("appActivity","com.walmart.android.app.main.MainActivity");
			driver =  new AndroidDriver(new URL("http://127.0.0.1:4723/wd/hub"),capabilities);
		}
		
		@Test
		
		public void operations(){
			driver.findElement(By.xpath("//android.widget.ImageButton[@content-desc='Navigate home']")).click();
		}

	}

**Comment by shabana-parveen** [comment]
@imurchie , I understand..

**Comment by MichaelVovk** [comment]
I have the same issue. Any solution?

**Comment by shabana-parveen** [comment]
Solution is please downgrade selenium version.

**Comment by krishna8181** [comment]
hi there 
please can you advise if we need to downgrade "selenium web driver java client" as well.
regards
krishna

**Comment by mattxie** [comment]
Hi there

Could anyone advise or point me to a online page on how to downgrade selenium version? I am using IntelliJ to develop a Appium app to test an IOS app and I am facing the same problem of this thread.

Thanks!

**Comment by mattxie** [comment]
I download selenium-remote-driver-3.3.1.jar from
   http://repo1.maven.org/maven2/org/seleniumhq/selenium/selenium-remote-driver/3.3.1/

   Renamed it to selenium-remote-driver-3.6.0.jar and repalced
   ~/.gradle/caches/modules-2/files-2.1/org.seleniumhq.selenium/selenium-remote-driver/3.6.0/9fc29afc07aa346ef0a84bdb4ddf68c50b84dfa9/selenium-remote-driver-3.6.0.jar

   Now it's working.

   But there must be a better way.

**Comment by shabana-parveen** [comment]
@polepalli-12 , I am using Maven project, please add dependencies for all jar file. Please share what kind of code you required from me? I'll create it for you to guide you. 
Thanks,

**Comment by InfantAnto** [comment]
I am also having the same issue with selenium latest versions.
The following events are not working with selenium latest versions 3.5.3, 3.6.0
click(), sendkeys(), getText()

But its working fine with version 3.3.1 . Is there any solutions for latest version?

**Comment by InfantAnto** [comment]
Hi All
I've fixed the issue by Upgrading the java-client version from 4.0.0 to 5.0.4 and the selenium latest version 3.6.0. Then click(), sendkeys() events worked for me.
```
<dependency>
            <groupId>io.appium</groupId>
            <artifactId>java-client</artifactId>
            <version>5.0.4</version>
</dependency>
```

**Comment by shabana-parveen** [comment]
@InfantAnto , Great, I'll check and let you know.

**Comment by shabana-parveen** [comment]
@InfantAnto , When can I suppose these versions as still its showing old versions ?please refer screenshot. 
![screen shot 2017-10-24 at 6 03 58 pm](https://user-images.githubusercontent.com/13347267/31943242-c1bc0ad6-b8e5-11e7-9fba-721f721d2518.png)

**Comment by antovrv** [comment]
@shabana-parveen 
I would suggest you to just remove both java-client and selenium-java from maven repository.
I got one new solution. Just have the java-client latest version in pom.xml.

```
<dependency>
            <groupId>io.appium</groupId>
            <artifactId>java-client</artifactId>
            <version>5.0.4</version>
</dependency>
```
Java-client automatically download the latest version of selenium-java(3.6.0) while mvn intsll. Kindly check and let me know.

Refer here. https://github.com/appium/java-client/blob/master/build.gradle#L57-L75

**Comment by InfantAnto** [comment]
@shabana-parveen 
I would suggest you to just remove both java-client and selenium-java from maven repository.
I got one new solution. Just have the java-client latest version in pom.xml.

```
<dependency>
            <groupId>io.appium</groupId>
            <artifactId>java-client</artifactId>
            <version>5.0.4</version>
</dependency>
```
Java-client automatically download the latest version of selenium-java(3.6.0) while mvn intsll. Kindly check and let me know.

Refer here. https://github.com/appium/java-client/blob/master/build.gradle#L57-L75

**Comment by shabana-parveen** [comment]
@InfantAnto Perfect, get latest selenium and java versions, Will let you know abt the NullPointerException..

**Comment by shabana-parveen** [comment]
@InfantAnto , I observed during testing that sometime click() operation works & sometime not for the same element.
Der might a be selenium problem as well, cause it happened usually with selenium, what to say. Please suggest.

I couldn't findout the root cause.

**Comment by InfantAnto** [comment]
Hi @shabana-parveen 
I had also faced the same issue before. Steps to verify the click is performed or not:

1) Enable the **_Pointer location_** in Developer Option Under Settings. Then you can see the click action is performed or not at the time of test running

2) Put Thread.sleep( 5000 ) - (for example 5 secs.) before and after the element click like below.

```
Thread.sleep(5000);
element.click();
Thread.sleep(5000);
```

That means we should give some time to click the element

(OR)

Now I am used like the below. It's works for me.

```
waitElementToBeClickableByLocator(driver, By.xpath("//something")).click();

**Function**:
public static WebElement waitElementToBeClickableByLocator(
            WebDriver driver, By by) {
        WebDriverWait wait = new WebDriverWait(driver, TEN_SECONDS);

        return wait.until(ExpectedConditions.elementToBeClickable(by));
    }

```

**Comment by TigMonkey** [comment]
@shabana-parveen, @InfantAnto  oh thank you, I resolved nullpointerException error.

**Comment by shabana-parveen** [comment]
@TigMonkey Great 👍

**Comment by InfantAnto** [comment]
@TigMonkey Sounds Good!

**Comment by andreyvasyliev** [comment]
@InfantAnto Thanks a lot! You solution really helps to fix nullpointerexception

**Comment by ravisankarb4u** [comment]
@InfantAnto, thank you very much. 
 
        <dependency>
            <groupId>io.appium</groupId>
            <artifactId>java-client</artifactId>
            <version>5.0.4</version>
        </dependency>

        
        <dependency>
            <groupId>org.seleniumhq.selenium</groupId>
            <artifactId>selenium-java</artifactId>
            <version>3.11.0</version>
        </dependency>

This combination resolved my nullpointerexception.

**Comment by archanapura** [comment]
Hi I tried with all above methods to solve null point exception but it didnt work me.
Its works in one of my system and fail in other.Can anyone help here?

**Comment by sqxz** [comment]
@archanapura you resolved the question? I can't fix it

**Comment by aman-daharwal** [comment]
Thanks guys the above discussion helped a lot and solved the problem.
The problem is while performing action on elements but after using below dependencies, the issue is resolved

`<dependency>
        <groupId>io.appium</groupId>
        <artifactId>java-client</artifactId>
        <version>5.0.4</version>
    </dependency>

    
    <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-java</artifactId>
        <version>3.11.0</version>
    </dependency>`

**Comment by ShankaKoththagoda** [comment]
Try @BeforeTest instead of @Before

**Comment by narendra-chandratre** [comment]
@InfantAnto @shabana-parveen I have tried solutions worked for you, but its failing for me. **NullPointer for PageFactory** class. 
Anyone has resolved this issue with **Appium 1.8.1 and Java-client 6.1.0**
